Proposal to use independent claims for each VC/VP proof type

Sakurann opened the pull request:
- Added four JWT Claims to represent W3C Verifiable Credentials objects
- Added vp_jwt, vp_ldp examples of SIOP response when W3C Verifiable Credentials objects are returned with JWTs 
- Added vp_jwt, vp_ldp examples of UserInfo response when W3C Verifiable Credentials objects are returned as JSON claims
- vc_jwt, vc_ldp option examples are missing
- took our vp_token related text
- edited the request part, moved it to the end for the following reasoning:
    - There are multiple candidates for requesting verifiable presentations and verifiable credentials using OpenID Connect flows: Edmund's Aggregated Claims draft, DIF Presentation Exchange, below draft, and probably others. This would be a natural next step after defining claims, and agreeing on the request syntax should be separate from agreeing on the usage of the above four claims.

Sakurann committed:
vp_jwt/vp_ldp proposal (Issue #16)

- Added four JWT Claims to represent W3C Verifiable Credentials objects
- Added vp_jwt, vp_ldp examples of SIOP response when W3C Verifiable Credentials objects are returned with JWTs 
- Added vp_jwt, vp_ldp examples of UserInfo response when W3C Verifiable Credentials objects are returned as JSON claims
- vc_jwt, vc_ldp options are missing

Sakurann committed:
vp_jwt/vp_ldp proposal (Issue #16)

- Added four JWT Claims to represent W3C Verifiable Credentials objects
- Added vp_jwt, vp_ldp examples of SIOP response when W3C Verifiable Credentials objects are returned with JWTs 
- Added vp_jwt, vp_ldp examples of UserInfo response when W3C Verifiable Credentials objects are returned as JSON claims
- vc_jwt, vc_ldp options are missing
I left out the request part for now for the discussion purposes

Sakurann committed:
Update README.md

Sakurann committed:
Merge pull request #1 from awoie/main

updating

Sakurann committed:
vp_jwt/vp_ldp proposal (Issue #16)

- Added four JWT Claims to represent W3C Verifiable Credentials objects
- Added vp_jwt, vp_ldp examples of SIOP response when W3C Verifiable Credentials objects are returned with JWTs 
- Added vp_jwt, vp_ldp examples of UserInfo response when W3C Verifiable Credentials objects are returned as JSON claims
- vc_jwt, vc_ldp option examples are missing
- took our vp_token related text
- edited the request part, moved it to the end for the following reasoning:
    - There are multiple candidates for requesting verifiable presentations and verifiable credentials using OpenID Connect flows: Edmund's Aggregated Claims draft, DIF Presentation Exchange, below draft, and probably others. This would be a natural next step after defining claims, and agreeing on the request syntax should be separate from agreeing on the usage of the above four claims.

Sakurann committed:
Update README.md

Sakurann committed:
Update README.md

tlodderstedt commented on README.md:
```diff
@@ -2,7 +2,7 @@
 ## Abstract
 
-This specification defines an extension of OpenID Connect to allow presentation of claims in the form of W3C Verifiable Credentials as part of the protocol flow in addition to claims provided in the `id_token` and/or via Userinfo responses.
+This specification defines additional parameters `vp_jwt`, `vp_ldp`, `vc_jwt`, `vc_ldp` to allow presentation of claims in the form of W3C Verifiable Credentials objects as part of the OpenID Connect protocol flow.
```
It does more than that as it includes syntax for requesting vcs and vps and will also contain processing rules. What's wrong with the existing text?

Sakurann replied:
I think syntax for requesting vcs and vps should be different from defining the claims. as iterated in the PR comment

tlodderstedt replied:
Sure. Requesting something is different from representing something. However, I intend to come up with at least one reasonable end 2 end solution. That's important for implementers and it is the basis for a stable and reasonable concept. Just discussion claims in isolation is not sufficient because those claims serve a purpose. There is not a single JWT claim so far defined in isolation. verified_claims, for example, is part of OpenID 4 Identity Assurance. It would never have gotten the maturity it has now without the context. Does this mean verified_claims cannot be used in other contexts? Of course not! We use it heavily in token introspection responses. 

I suggest to first of all discuss this topic and then propose potential changes to our draft in a separate PR. So far, we have discussed to change the representation of the VCs/VPs from vp_token and id token as VP to universal claims in ID token (and other artefacts). We agreed to continue this discussion and you took the task to make a PR as basis for further discussions. The current PRs goes way beyond what we have discussed. It especially proposes removal of text that I consider important and have put significant work in. I don't agree with those changes.

Sakurann replied:
> The current PRs goes way beyond what we have discussed. It especially proposes removal of text that I consider important and have put significant work in. I don't agree with those changes.

yes, I understand. It felt wrong when doing this PR. Let me revert

tlodderstedt commented on README.md:
```diff
@@ -68,465 +69,304 @@ This approach dramatically reduces latency and reduces load on the OP's servers.
-* ID Token as Verififiable Presentation: An ID Token may contain a claim `vp` or `vc` as defined in [JWT proof format](https://www.w3.org/TR/vc-data-model/#json-web-token), i.e. it is a valid OpenID Connect ID Token and a VC or VP at the same time. Consequently, this mechanism utilizes (and supports) the external JWT proof format only. 
-* VP Token: a Verifiable Presentation is provided in a separate artifact designated as "VP Token". Such a token is provided to the RP in addition to an `id_token` in the `vp_token` parameter. VP Tokens support Verifiable Presentations in JSON-LD as well JWT format including all respective proof formats. They also allow to sign ID Token and Verifiable Presentation with different key. 
-* VC Token: a Verifiable Credential is provided in a separate artifact designated as "VC Token". Such a token is provided to the RP in addition to an `id_token` in the `vc_token` parameter. VC Tokens support Verifiable Presentations in JSON-LD as well JWT format including all respective proof formats.
+* JWTs such as ID Tokens may contain a claim `vp_jwt`, `vp_ldp`, `vc_jwt`, or `vc_ldp`.
```
I don't get the value of the differentiation between JWTs and JSON claim set. These claims can be used in any JSON based object, right? The payload of a JWT is a JSON claim set as well, isn’t it?

Sakurann replied:
The UserInfo Endpoint uses JWT claims in JSON, but without the response being a JWT. It's the body of a JWT, without the crypto parts.

tlodderstedt replied:
I think we are saying the same. I would suggest to come up with more generic text like we have in eKYC. 

```
A verified_claims element can be added to a OpenID Connect UserInfo response or an ID Token.

OAuth Authorization Servers can add verified_claims to access tokens in JWT format or Token Introspection responses, either in plain JSON or JWT-protected format.

An OP or AS MAY also include verified_claims in the beforementioned assertions as aggregated or distributed claims (see Section 5.6.2 of the OpenID Connect specification [OpenID]).

```
PS: a UserInfo response can also be a JWT if it is signed and/or encrypted ;-).

Sakurann replied:
tried to incorporate proposed generic text :)

tlodderstedt commented on README.md:
```diff
@@ -68,465 +69,304 @@ This approach dramatically reduces latency and reduces load on the OP's servers.
+* Sets of JSON claims that can be returned from the endpoints such as UserInfo endpoint may contain a claim `vp_jwt`, `vp_ldp`, `vc_jwt`, or `vc_ldp`.
 
-This table shows the different combinations of signatures on id token, VC, and VP and how the binding of the VC or VP with the holder is validated by the RP.
+Note that OP would first encode VPs/VCs using the rules defined in the Verifiable Credential specification either in JWT format or JSON-LD format, before passing encoded VPs/VCs as `vp_jwt`, `vp_ldp`, `vc_jwt`, or `vc_ldp` parameters as JWT claims or as sets of JSON claims.
```
feels better placed in a section about request processing

tlodderstedt commented on README.md:
```diff
@@ -68,465 +69,304 @@ This approach dramatically reduces latency and reduces load on the OP's servers.
+This specification defines the following parameters to pass Verifiable Presentations or Verifiable Credentials signed as JWTs or using Linked Data Proofs:
 
-### vp in id_token
+- vc_jwt:  A claim whose value is a W3C Verifiable Credential object using the JWT representation, which is a JSON string.  The claim’s value may also be an array of W3C Verifiable Credential objects using the JWT representation if the use case calls for multiple JWT VCs.
```
I don't particularly like this polymorphism.

Sakurann replied:
you fundamentally object to the choice of the four claims?

tlodderstedt replied:
This comment is about the polymorphic definition of the claims as object or array. It is inline with the way aud is defined, but it creates unnecessary complexity where an array is sufficient.

My next comment raises the question whether the design approach is the best solution and proposes an alternative. There are pros and cons of both proposals that we should discuss and assess based on use cases and other requirements informing this discussion.

awoie replied:
I agree that the polymorphism also caused some issues with `aud` implementations in the past. I'm happy to just say it is an array.

tlodderstedt commented on README.md:
```diff
@@ -68,465 +69,304 @@ This approach dramatically reduces latency and reduces load on the OP's servers.
-A RP requests a Verifiable Presentation using the `claims` parameter. 
+W3C Verifiable Credentials specification defines two kinds of objects – Verifiable Credentials and Verifiable Presentations, and it also orthogonally defines two proof formats of these objects – JWT and Linked Data Proofs. Thus, there are four data types that different use cases could utilize.
+ 
+This specification defines the following parameters to pass Verifiable Presentations or Verifiable Credentials signed as JWTs or using Linked Data Proofs:
```
I'm not sure whether we should define 4 different claims. I feel a single claim with metadata is better suited. 
Something like this:
```json
   "verifiable_presentation":{
      "format":"jwt",
      "proof":"external",
      "presentation":"ewogICAgImlzcyI6Imh0dHBzOi8vYm9vay5pdHNvdXJ3ZWIub...IH0="
   }
```
This gives implementors the ability to support further formats, like CBOR or AnonCreds. I would also envision implementations to add further data to the structure as needed, e.g. the metadata we talked about yesterday for carrying additional metadata for hash-based VCs in JWT-format to allow for selective disclosure.

awoie replied:
If we used the metadata approach, then we should really think whether it makes sense to borrow the `presentation_submission` property from DIF PE:

```  
"presentation_submission": {
    "id": "a30e3b91-fb77-4d22-95fa-871689c322e2",
    "definition_id": "32f54163-7166-48f1-93d8-ff217bdb0653",
    "descriptor_map": [
      {
        "id": "banking_input_2",
        "format": "jwt_vc",
        "path": "$.verifiableCredential[0]"
      },
      {
        "id": "employment_input",
        "format": "ldp_vc",
        "path": "$.verifiableCredential[1]"
      },
      {
        "id": "citizenship_input_1",
        "format": "ldp_vc",
        "path": "$.verifiableCredential[2]"
      }
    ]
  }
```

The `path` would just point to the `verifiable_presentation` claim in the `id_token`.

The only problem is that `definition_id` is mandatory and refers to the `presentation_definition` in the request. A minimal `presentation_definition` looks as follows:

```
"presentation_definition": {
    "id": "32f54163-7166-48f1-93d8-ff217bdb0653",
    "input_descriptors": [
      {
        "id": "wa_driver_license",
        "schema": [{
            "uri": "https://licenses.example.com/business-license.json"
        }]
      }
    ]
}
```

Sakurann replied:
I am in favor of the array approach with several claims that address most needed representations, which would be more simple. Since I do not think we expect wide adoption of new proof formats other than JSON/JSON-LD in the mid-term future, four claims should fulfill the implemetations' needs. 

Having said that, I think we agreed to start a separate in-depth document comparing the two approaches - four claims vs. one claim with metadata.

tlodderstedt commented on README.md:
```diff
@@ -68,465 +69,304 @@ This approach dramatically reduces latency and reduces load on the OP's servers.
+  "typ": "JWT",
+  "alg": "ES256K"
+}.{
+   "iss":"https://book.itsourweb.org:3000/wallet/wallet.html",
```
is this a valid SIOP iss claim?

awoie replied:
self-issued.me is no longer a SIOP requirement?

Sakurann replied:
it is, I edited this out.

tlodderstedt commented on README.md:
```diff
@@ -68,465 +69,304 @@ This approach dramatically reduces latency and reduces load on the OP's servers.
+# Appendix?
 
-#### Claims parameter
+There are multiple candidates for requesting verifiable presentations and verifiable credentials using OpenID Connect flows: Edmund's Aggregated Claims draft, DIF Presentation Exchange, below draft, and probably others. This would be a natural next step after defining claims, and agreeing on the request syntax should be separate from agreeing on the usage of the above four claims.
```
Please reintegrate the request section and the request examples in its previous places. I would then adopt them to fit the new claims.

Sakurann replied:
yes, updated

tlodderstedt reviewed (changes requested):
Please see my comments. Additionally, it seems significant parts of the main revision re are still missing. Please integrate again.

Sakurann committed:
Reflected comments

- Re-added text and examples of the request
- Rdited text

awoie commented on README.md:
```diff
@@ -66,121 +67,80 @@ This approach dramatically reduces latency and reduces load on the OP's servers.
-| Holder of the VC | bearer credential or same did in `sub` and credential | vp signed by holder | bearer credential or same did in `sub` and credential + `vc_hash` | VP signed by holder + `vp_hash` 
-| Other entity (e.g. OP)| bearer credential | n/a | bearer credential  | VP signed by holder + `vp_hash`
 
-## Requesting Verifiable Presentations
```
I think the claims are a good starting point but this particular spec should also talk about the protocol level, i.e., requesting Verifiable Presentations. Note, the DIF Presentation Exchange spec only contains examples for OIDC but there is no normative reference. Someone and I propose we do this in the course of this spec, should create such a normative reference by utilizing the proposed claims.

Sakurann replied:
yes, agreed. per the call we just had, I think the next step is to show how VP/VCs will be requested using 
- OIDC claims parameter
- DIF Presentation Exchange
- Aggregation claims draft
- OIDC ekyc-ida?
- Credential Provider draft?

tlodderstedt replied:
Most of the example you gave us the claims parameter in different flavours.

The aggregation claims draft doesn’t cover VP/VC. It uses the claims parameter to let an OP request claim sets at an aggregated claims provider for further use towards a RP as aggregated claims. 

OIDC ekyc-ida uses the claims parameter to request verified_claims. That's a different topic but we should borrow the pattern. We perhaps can also utilize eKYC syntax, depending which way we go re representation and what the intended scope is.

Credential Provider draft uses a scope to request issuance of credentials and is silent on how the OP (credential provider) determines the credential types. From the conversations with Adam und Tobias we learned the idea is to a single OP per credential type to issue. I think this model won't scale. I think credential representation and how to request credentials should be aligned with this draft. 

I think we've got two options: claims or DIF PE

awoie reviewed (changes requested):
i guess we need to reiterate on the query request format a bit.

tlodderstedt commented on README.md:
```diff
@@ -66,121 +67,80 @@ This approach dramatically reduces latency and reduces load on the OP's servers.
+| Proof format on the object| JWT | JWT | LD-Proof | LD-Proof
 
-### vp_token
+There are multiple candidates for requesting verifiable presentations and verifiable credentials using OpenID Connect flows: Edmund's Aggregated Claims draft, DIF Presentation Exchange, below draft, and probably others. This would be a natural next step after defining claims, and agreeing on the request syntax should be separate from agreeing on the usage of the above four claims.
```
```suggestion
The next section illustrates how the `claims` parameter can be used for requesting verified presentations. It serves as a starting point to drive discussion about this aspect. There are other candidate approaches for this purpose. They will be evaluated as this draft evolves. 
```

tlodderstedt commented on README.md:
```diff
@@ -66,121 +67,80 @@ This approach dramatically reduces latency and reduces load on the OP's servers.
+A RP requests a Verifiable Presentation using the `claims` parameter. 
 
-* `claims` - An object determining the claims the RP wants to obtain using the same notation as used underneath `id_token`. 
+### Verifiable Presentation object in id_token
```
the section is still just a small part of what exists in the main revision. 

For example, it misses the ability to request a subset of claims.

tlodderstedt commented on README.md:
```diff
@@ -205,243 +165,143 @@ In this case, the RP asks the OP to provide a VC of a certain type.
+### ID Token with Verifiable Presentation signed using Linked Data Proofs
 
-### Authentication Response 
+Below is a non-normative example of ID Token that includes `vp_ldp` claim.
```
request example is missing

Sakurann replied:
added

tlodderstedt commented on README.md:
```diff
@@ -461,8 +321,7 @@ This section illustrates the protocol flow for the case of communication using f
-    "vp_token": {
-      "format": "json-ld",
+    "vp_ldp": {
       "claims":
```
Ambiguous request - it shall either put the vp_ldp into an "id_token" or "userinfo" context.

Sakurann replied:
requesting "vp_ldp" using "userinfo". Added reqesting "auth_time" using "id_token" to illustrate  how these can work together.

tlodderstedt reviewed (changes requested):
request syntax definition is incomplete and auth code example needs fix

tlodderstedt commented:
That‘s all claims parameter based. The only exception is PE.

I created a ticket
https://github.com/awoie/vp-token-spec/issues/22

Kristina ***@***.***> schrieb am Fr. 9. Apr. 2021 um 18:33:

> *@Sakurann* commented on this pull request.
> ------------------------------
>
> In README.md
> <https://github.com/awoie/vp-token-spec/pull/20#discussion_r610762790>:
>
> >
> -## Requesting Verifiable Presentations
>
> yes, agreed. per the call we just had, I think the next step is to show
> how VP/VCs will be requested using
>
>    - OIDC claims parameter
>    - DIF Presentation Exchange
>    - Aggregation claims draft
>    - OIDC ekyc-ida?
>    - Credential Provider draft?
>
> —
> You are receiving this because your review was requested.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/awoie/vp-token-spec/pull/20#discussion_r610762790>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACQFJT6WWGVSOFBE7HCFY3TH4T45ANCNFSM42UGEJKA>
> .
>

Sakurann committed:
update requesting VP section intro

Co-authored-by: Torsten Lodderstedt <torsten@lodderstedt.net>

Sakurann committed:
Adding examples

- Adding missing examples
- Clarifying the structure

Sakurann commented on README.md:
```diff
@@ -198,252 +165,167 @@ The following is a non-normative example of how an RP would use the `claims` par
-   },
    "nonce":"960848874",
+   "vp_jwt":[
+            "ewogICAgImlzcyI6Imh0dHBzOi8vYm9vay5pdHNvdXJ3ZWIub...IH0="
```
I am not sure Aggregated Claims syntax really makes sense. Claims would have to be requested using their pre-agreed names in the request which does not really work with VPs in JSON-LD as it dilutes the purpose of @context. 

Aggregated Claims response example:
```json
{
  "_claim_names": {
       "address": "src1",
       "phone_number": "src1"
     },
   "_claim_sources": {
     "src1": {"vp_jwt": "ewogICAgImlzcyI6Imh0dHBzOi8vYm9vay5pdHNvdXJ3ZWIub...IH0="}
   }  
```

Sakurann replied:
or maybe it does, OP would just take the claim name out of the schema

tlodderstedt replied:
the standard aggregated claims syntax does only allow you to refer to the top level claims provided in the aggregated claim set. In your example, this would be:

```json
{
  "_claim_names": {
       "vp_jwt": "src1"
     },
   "_claim_sources": {
     "src1": {"vp_jwt": "ewogICAgImlzcyI6Imh0dHBzOi8vYm9vay5pdHNvdXJ3ZWIub...IH0="}
   }
```
telling the RP where to look for the `vp_jwt` claim. 

In order to support multiple occurences (multiple VPs), one could adopt the syntax extension we did in eKYC (https://openid.net/specs/openid-connect-4-identity-assurance-1_0.html#name-verified_claims-delivery). We also allow for an array of sources.

tlodderstedt commented on README.md:
```diff
@@ -198,252 +165,167 @@ The following is a non-normative example of how an RP would use the `claims` par
+### Authorization Code Flow with Verifiable Presentation in ID Token
 
-This section illustrates the protocol flow for the case of communication using frontchannel and backchannel (utilizing the authorization code flow).
+Below are the examples when verifiable presentation is requested and returned from user_info endpoint responses as part of OpenID Connect Authorization Code Flow. ID Token contains a `vp_jwt` or `vp_ldp` element with the Verifiable Presentation data. It can also contain `vc_jwt` or `vc_ldp` element with the Verifiable Credential data. 
```
I suggest to start with VPs in ID token and then give the userinfo response example. The current text might create the false impression code means VPs are delivered via userinfo. I think, the majority of RPs will use the ID Token for that purpose. Our partners (at least) do it with standard OIDC claims as way.

tlodderstedt commented on README.md:
```diff
@@ -457,31 +339,65 @@ This section illustrates the protocol flow for the case of communication using f
+   "userinfo":
+    {
+      "vp_jwt": {
+        "claims":
```
this syntax is not covered by the syntax definition. I suggest to use the example fron above with the credential_types element. We can extend the query syntax once we have come to a consensus re claims vs container vs ADS vs ...

Sakurann replied:
for `vp_jwt` changed to response_type; for `vp_ldp`, it is response_type + claims

tlodderstedt commented on README.md:
```diff
@@ -457,31 +339,65 @@ This section illustrates the protocol flow for the case of communication using f
+   "userinfo":
+    {
+      "vp_ldp": {
+        "claims":
```
see above

tlodderstedt commented on README.md:
````diff
@@ -457,31 +339,65 @@ This section illustrates the protocol flow for the case of communication using f
 ```
 
-### Authentication Response
+#### Claims parameter (vp_ldp)
````
I suggest to show associated claims parameter example and token response example in a row.

vp_jwt
- claims 
- token response
- vp

vp_ldp
- claims
- token response
- vp

Sakurann replied:
vp is quite different ,but in claims, vp_jwt simply changes to vp_ldp and token response is the same

Sakurann replied:
restructured as above

tlodderstedt commented on README.md:
````diff
@@ -516,17 +421,94 @@ HTTP/1.1 302 Found
+```
 
-# vc_token encoding options
+## VC encoding options
````
I don't think we need this section any longer

tlodderstedt commented on README.md:
```diff
@@ -66,121 +71,83 @@ This approach dramatically reduces latency and reduces load on the OP's servers.
-* ID Token as Verififiable Presentation: An ID Token may contain a claim `vp` or `vc` as defined in [JWT proof format](https://www.w3.org/TR/vc-data-model/#json-web-token), i.e. it is a valid OpenID Connect ID Token and a VC or VP at the same time. Consequently, this mechanism utilizes (and supports) the external JWT proof format only. 
-* VP Token: a Verifiable Presentation is provided in a separate artifact designated as "VP Token". Such a token is provided to the RP in addition to an `id_token` in the `vp_token` parameter. VP Tokens support Verifiable Presentations in JSON-LD as well JWT format including all respective proof formats. They also allow to sign ID Token and Verifiable Presentation with different key. 
-* VC Token: a Verifiable Credential is provided in a separate artifact designated as "VC Token". Such a token is provided to the RP in addition to an `id_token` in the `vc_token` parameter. VC Tokens support Verifiable Presentations in JSON-LD as well JWT format including all respective proof formats.
+OAuth Authorization Servers can add `vp_jwt`, `vp_ldp`, `vc_jwt`, or `vc_ldp` claims to ID tokens in JWT format or UserInfo responses either in plain JSON or JWT-protected format.
```
You mean OpenID Connect OPs? I'm asking since OAuth Authorization Servers issue access tokens and may expose token introspection but don't issue ID Tokens or have a userinfo endpoint.

Sakurann replied:
good catch, thank you. I meant OPs.

tlodderstedt commented on README.md:
```diff
@@ -66,121 +71,83 @@ This approach dramatically reduces latency and reduces load on the OP's servers.
 ## Overview
 
-This specifications introduces the following mechanisms to provide VCs and VPs to RPs:
+Verifiable Credentials and Verifiable Presentations can be added to a OpenID Connect UserInfo response or an ID Token.
```
I don't fully get why we need the following 3 sentences since the real definition is given in "ID Token Extensions".

I suggest to remove the additional heading "ID Token Extensions" and move (a simplified version of) the first 3 sentences to the end of the section. 

Text could be: 

"Those claims can be added to ID Tokens, Userinfo responses as well as Access Tokens and Introspection response. "

tlodderstedt commented on README.md:
```diff
@@ -66,121 +71,83 @@ This approach dramatically reduces latency and reduces load on the OP's servers.
-`proof_format`
-[TBD]
+`credential_types`
+A string array containing a list of VC credential types the RP asks for. The OP shall respond with a presentation containing one credential of one of the listed types. 
```
definition for "userinfo" is missing

Sakurann replied:
as in [https://openid.net/specs/openid-connect-core-1_0.html#ClaimsParameter](https://openid.net/specs/openid-connect-core-1_0.html#ClaimsParameter)?

Sakurann replied:
I think it should be "The OP shall respond with a presentation containing credentials of the listed types."

As the text states now, it is misleading that array is how a client asks for any one​ of these types to be satisfied (not all​ of them)

tlodderstedt reviewed (changes requested):
I added a couple of comments.

Sakurann committed:
update examples

- changed request syntax (for vp_jwt response_type; for vp_ldp, it is response_type + claims)
- restructured to request and response for vp_jwt, request and response for vp_ldp
- took out ## VC encoding options section..
- edited Overview section
- added one sentence about DIF PE
